lsm-perf.py

```diff
@@ -0,0 +1,180 @@
+    print_eta(name, info='connecting')
+
+    with VM(kernel_path, img_path, keyfile) as vm:
+        vm.scp_to(workload_path, ON_VM_WORKLOAD_PATH)
+        work_cmd = vm.ssh[ON_VM_WORKLOAD_PATH]
+
+        print_eta(name, info='Running warm up')
+        for _ in range(WARMUP_RUNS):
+            work_cmd()
+
+        for i in range(REPETITIONS_PER_ROUND):
+            results.append(int(work_cmd().strip()))
+            percentage = (i + 1) * 100 / REPETITIONS_PER_ROUND
+            print_eta(name, info='%d%%' % percentage)
+
+        vm.ssh.path(ON_VM_WORKLOAD_PATH).delete()
+
+    stats = ('\taverage=%d, stdev=%d' %
+             (statistics.mean(results), statistics.stdev(results)))
+    print_eta(name, info=stats)
+    print()
+    return results
+
+
+class VM:
+    """Represents a qemu VM with an SSH connection.
+    To use with the `with` construct
+    """
+    def __init__(self, kernel_path, img_path, keyfile):
+        """Starts the qemu VM"""
+        qemu_args = construct_qemu_args(kernel_path, img_path)
+        self.process = local['qemu-system-x86_64'].popen(qemu_args)
+        self.ssh = None
+        self.key = keyfile
+
+    def __enter__(self):
+        """Initialize ssh connection"""
+        c = 5
+        while self.ssh is None:
+            time.sleep(1)
+            try:
+                self.ssh = SshMachine(
+                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
+            except (EOFError, plumbum.machines.session.SSHCommsError) as e:
+                c -= 1
+                if c == 0:
+                    raise e
+        return self
+
+    def __exit__(self, type, value, traceback):
+        """Stops the SSH connection and the VM"""
+        if self.ssh is not None:
+            self.ssh.close()
+            self.ssh = None
+        self.process.stdin.write(QEMU_EXIT_CMD)
+        self.process.terminate()
+        time.sleep(0.5)
+
+    def scp_to(self, src_local, dst_remote):
+        """Send a file from the host to the VM"""
+        assert self.ssh is not None
+        fro = local.path(src_local)
+        to = self.ssh.path(dst_remote)
+        plumbum.path.utils.copy(fro, to)
+
+
+def construct_qemu_args(kernel_path, img_path):
+    """Qemu arguments similar to what `vm start` produces"""
+    return [
+        '-nographic',
+        '-s',
+        '-machine', 'accel=kvm',
+        '-cpu', 'host',
+        '-device', 'e1000,netdev=net0',
+        '-netdev', 'user,id=net0,hostfwd=tcp::5555-:22',
+        '-append', 'console=ttyS0,115200 root=/dev/sda rw nokaslr',
+        '-smp', '4',
+        '-m', '4G',
+        '-drive', 'if=none,id=hd,file=%s,format=raw' % img_path,
+        '-device', 'virtio-scsi-pci,id=scsi',
+        '-device', 'scsi-hd,drive=hd',
+        '-device', 'virtio-rng-pci,max-bytes=1024,period=1000',
+        '-qmp', 'tcp:localhost:4444,server,nowait',
+        '-serial', 'mon:stdio',
+        '-kernel', '%s' % kernel_path,
+        '-name', 'lsm_perf_vm,debug-threads=on'
```

Comment:
The args constructed here are tightly coupled with the `VM` class - e.g. if you change the `5555` it won't be able to connect any more. So I think it should be a method of that class. Those couplings would also ideally be expressed directly in the code: instead of having a `5555` in this string and then a 5555 in the code, we should have a single variable set to 5555.

Reply:
Ok I updated it, it makes more sense indeed. The port might also become a parameter in the future.